examples/cluster/plot_kmeans_silhouette_analysis.py

```diff
@@ -0,0 +1,139 @@
+    # clusters
+    silhouette_avg = silhouette_score(X, cluster_labels)
+    print("For n_clusters =", n_clusters,
+          "The average silhouette_score is :", silhouette_avg)
+
+    # Compute the silhouette scores for each sample
+    sample_silhouette_values = silhouette_samples(X, cluster_labels)
+
+    # This will hold the silhouette coefficient of all the clusters separated
+    # by 0 samples.
+    sorted_clustered_silhouette_values = []
+
+    y_lower = 10
+    for i in np.unique(cluster_labels):
+        # Aggregate the silhouette scores for samples belonging to
+        # cluster i, and sort them
+        ith_cluster_silhouette_values = \
+            sample_silhouette_values[cluster_labels == i]
+
+        ith_cluster_silhouette_values.sort()
+
+        size_cluster_i = ith_cluster_silhouette_values.shape[0] 
+        y_upper = y_lower + size_cluster_i
+
+        color = cm.spectral(float(i) / n_clusters, 1)
+        ax1.fill_betweenx(np.arange(y_lower, y_upper),
+                          0, ith_cluster_silhouette_values,
+                          facecolor=color, edgecolor=color)
+
+        # Label the silhouette plots with their cluster numbers at the middle
+        ax1.text(-0.05, y_lower + 0.5 * size_cluster_i, str(i))
+
+        # Compute the new y_lower for next plot
+        y_lower = y_upper + 10  # 10 for the 0 samples
+
+        # Store the sorted values separating them with 10 0 samples
+        sorted_clustered_silhouette_values += \
+            ith_cluster_silhouette_values.tolist() + [0] * 10
+
+    ax1.set_title("The silhouette plot for the various clusters.")
+    ax1.set_xlabel("The silhouette coefficient values")
+    ax1.set_ylabel("Cluster label")
+    
+    # The vertical line for average silhoutte score of all the values
+    ax1.axvline(x = silhouette_avg, color = "red", linestyle = "--")
+
+    ax1.set_yticks([])  # Clear the yaxis labels / ticks
+    ax1.set_xticks([-0.1, 0, 0.2, 0.4, 0.6, 0.8, 1])
+
+    for k in range(len(X)):
+        color = cm.spectral(float(cluster_labels[k]) / n_clusters, 1)
+        ax2.scatter(X[k, 0], X[k, 1], marker='.', color=color)
+
+    # 2nd Plot showing the actual clusters formed
+    for c in clusterer.cluster_centers_:
+        # Use the clusterer to know to which cluster number the current center
+        # c belongs to
+        i = clusterer.predict(c)[0]
+
+        # Scatter plot of all the ith cluster members
+        cluster_members = X[cluster_labels == i]
+        color = cm.spectral(float(i) / n_clusters, 1)
+        ax2.scatter(cluster_members[:, 0], cluster_members[:, 1], 
+                    marker='.', color=color)
+ 
+        # Label the cluster centers with the cluster number for identification 
+        # and study of the corresponding silhouette plot.
+        ax2.scatter(c[0], c[1], marker='o', c="white", alpha=1, s=200)
+        ax2.scatter(c[0], c[1], marker="$%d$" % i, alpha=1, s=50)
```

Comment:
I'm new to this... Is there a better way to plot the clusters and label the centers?